lib/requests/request_executor.rb

```diff
@@ -97,6 +230,280 @@ def dispose
       cancel_failing_nodes_timers
```

Comment:
There's more happening in the JVM version here in `dispose()` (AKA `close()` as it's called there)

Reply:
I will check this along with introducing semaphores.